keps/sig-cli/kustomize-ordering.md

```diff
@@ -0,0 +1,179 @@
+sorting by Resource Type is insufficient.
+
+### Goals
+
+- Provide the ability for users to override the order that Kustomize emits Resources
+  - Used by `kubectl apply` or order Resource operations
+  - Used to override creation order for meta-Resources such as Namespaces
+
+### Non-Goals
+
+- Ensure certain Resources are *Settled* or *Ready* before other Resources
+  - Ensure dependencies between Workloads
+
+## Proposal
+
+Provide a simple mechanism allowing users to override the order that
+Resource operations are Applied.  Add a new field `sortOrder` that is
+a list of `ResourceSelector`s which match Resources based
+off their annotations.
+
+```go
+
+type Kustomization struct {
+	// ...
+	
+	// sortOrder defines a precedence for ordering Resources.  Resources
+	// will be sorted in the order the match a ResourecSelector before
+	// they are emitted.
+	SortOrder []ResourceSelector `json:"sortOrder"`
+
+	// ...
+}
+
+// ResourceSelector selects Resources that it matches
+type ResourceSelector struct {
+	// matchAnnotations is a map of {key,value} pairs.  A Resource matches if it has
+	// *all* of the annotations in matchAnnotations appear in the Resource metaData.
+	// Null and empty values are not allowed.
+	// +optional
+	MatchAnnotations map[string]string `json:"matchAnnotations,omitempty"`
+	
+	// TODO: Consider adding field: MatchExpressions []AnnotationSelectorRequirement 
+}
+
+```
+Example:
+
+```
+sortOrder:
+- matchAnnotations:
+    some-annotation-name-1: some-annotation-value-1
+    some-annotation-name-a: some-label-value-a
+- matchAnnotations:
+    some-annotation-name-1: some-annotation-value-1
+- matchAnnotations:
+    some-annotation-name-2: some-annotation-value-2
+```
+
+The explicit user defined ordering using annotations will take precedence over the type based
+orderings.  Types would be used as a fallback sorting function amongst Resource with equal
+precedence in the explicit ordering.
+
+- Resources annotated with `some-annotation-name-1: some-annotation-value-1` *and* annotated
+  with `some-annotation-name-a: some-annotation-value-a` are first
+  - These Resources are sorted by type
+- Resources annotated with `some-annotation-name-1=some-annotation-value-1`
+  (without `some-annotation-name-a: some-annotation-value-a`) appear second
+  - These Resources are sorted by type
+- Resources annotated with `some-annotation-name-2=some-annotation-value-2` appear third
+  - These Resources are sorted by type
+- Resources not matching any annotation are last
+  - These Resources are sorted by type
+
+Resources matching multiple orderings (e.g. have multiple matching annotations) appear
+in the position matching the earliest label / annotation.
+
+**Note:** Throw an error if there is a selector that selects a superset of another selector 
+and it appears first.  e.g. this should throw an error because the first selector will match
+everything that the second selector does, and it will have no effect.
+
+```
+sortOrder:
```

Comment:
How about generating the output in the same order as they are listed in `resources`. If one resource file containing multiple resource configs, keep their original order?  Moreover, keep the order of `bases`.
In `Kustomization.yaml`, we can add one field to switch to this behavior:
```
keepOriginalOrder: true
```